util.go

```diff
@@ -50,17 +69,96 @@ func (info *WriteFileInfo) Validate() error {
 	if info.Perm == 0 {
 		info.Perm = 0666
 	}
+
+	catcher.Add(info.validateContent())
+
 	return catcher.Resolve()
 }
 
 // DoWrite writes the data to the given path, creating the directory hierarchy as
-// needed.
-func (info WriteFileInfo) DoWrite() error {
+// needed and the file if it does not exist yet.
+func (info *WriteFileInfo) DoWrite() error {
 	if err := makeEnclosingDirectories(filepath.Dir(info.Path)); err != nil {
 		return errors.Wrap(err, "problem making enclosing directories")
 	}
 
-	return errors.Wrap(ioutil.WriteFile(info.Path, info.Data, info.Perm), "problem writing file")
+	openFlags := os.O_RDWR | os.O_CREATE
+	if info.Append {
+		openFlags |= os.O_APPEND
+	} else {
+		openFlags |= os.O_TRUNC
+	}
+
+	file, err := os.OpenFile(info.Path, openFlags, 0666)
+	if err != nil {
+		return errors.Wrapf(err, "error opening file %s", info.Path)
+	}
+
+	catcher := grip.NewBasicCatcher()
+
+	reader, err := info.ContentReader()
+	if err != nil {
+		catcher.Wrap(file.Close(), "error closing file")
+		catcher.Wrap(err, "error getting file content as bytes")
+		return catcher.Resolve()
+	}
+
+	const mb = 1024 * 1024
+	buf := make([]byte, mb)
+	for n, err := reader.Read(buf); n > 0; n, err = reader.Read(buf) {
+		if err != nil && err != io.EOF {
+			catcher.Wrap(file.Close(), "error closing file")
+			catcher.Wrap(err, "error reading content to write to file")
+			return catcher.Resolve()
+		}
+		if _, err := file.Write(buf[:n]); err != nil {
+			catcher.Wrap(file.Close(), "error closing file")
+			catcher.Wrap(err, "error writing content to file")
+			return catcher.Resolve()
+		}
+	}
+
+	return errors.Wrap(file.Close(), "error closing file")
+}
+
+// SetPerm sets the file permissions on the file. This should be called after
+// DoWrite. If no file exists at (WriteFileInfo).Path, it will error.
+func (info *WriteFileInfo) SetPerm() error {
+	return errors.Wrap(os.Chmod(info.Path, info.Perm), "error setting permissions")
+}
+
+// ContentBytes returns the contents to be written to the file as a byte slice.
+func (info *WriteFileInfo) ContentBytes() ([]byte, error) {
+	if err := info.validateContent(); err != nil {
+		return nil, errors.Wrap(err, "could not validate file content source")
+	}
+
+	if info.Reader != nil {
+		content, err := ioutil.ReadAll(info.Reader)
```

Comment:
bufio rather than readall?

Reply:
I changed `ContentBytes()` to be unexported and use it for reading from the reader in buffered chunks.